File: src/cstlib/memory.py
"""Durable semantic memory with a dependency-free hashed embedding fallback."""
from __future__ import annotations

import hashlib
import json
import math
import re
import time
import uuid
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Callable, Iterable

Embedding = list[float]
Embedder = Callable[[str], Iterable[float]]
_TOKEN = re.compile(r"[A-Za-z0-9_']+")


def hashed_embedding(text: str, dimension: int = 128) -> Embedding:
    vector = [0.0] * dimension
    for token in _TOKEN.findall(text.lower()):
        digest = hashlib.blake2b(token.encode("utf-8"), digest_size=16).digest()
        index = int.from_bytes(digest[:8], "big") % dimension
        sign = 1.0 if digest[8] & 1 else -1.0
        vector[index] += sign
    norm = math.sqrt(sum(v * v for v in vector))
    return [v / norm for v in vector] if norm else vector


def cosine(a: Iterable[float], b: Iterable[float]) -> float:
    av, bv = list(a), list(b)
    if len(av) != len(bv):
        raise ValueError("embedding dimensions differ")
    denom = math.sqrt(sum(x * x for x in av)) * math.sqrt(sum(y * y for y in bv))
    return sum(x * y for x, y in zip(av, bv)) / denom if denom else 0.0

@dataclass(slots=True)
class MemoryRecord:
    id: str
    text: str
    timestamp: float
    salience: float = 0.5
    confidence: float = 1.0
    metadata: dict[str, object] | None = None

class SemanticMemory:
    def __init__(self, path: str | Path | None = None, *, embedder: Embedder | None = None, similarity_weight: float = 0.70, recency_weight: float = 0.10, salience_weight: float = 0.15, confidence_weight: float = 0.05) -> None:
        self.path = Path(path) if path else None
        self.embedder = embedder or hashed_embedding
        self.weights = (similarity_weight, recency_weight, salience_weight, confidence_weight)
        self.records: list[MemoryRecord] = []
        self._embeddings: dict[str, Embedding] = {}
        if self.path and self.path.exists():
            self._load()

    def _load(self) -> None:
        assert self.path is not None
        for line in self.path.read_text(encoding="utf-8").splitlines():
            if not line.strip():
                continue
            record = MemoryRecord(**json.loads(line))
            self.records.append(record)
            self._embeddings[record.id] = list(self.embedder(record.text))

    def store(self, text: str, *, salience: float = 0.5, confidence: float = 1.0, metadata: dict[str, object] | None = None) -> MemoryRecord:
        if not text.strip():
            raise ValueError("memory text cannot be empty")
        record = MemoryRecord(uuid.uuid4().hex, text, time.time(), max(0.0, min(1.0, salience)), max(0.0, min(1.0, confidence)), metadata)
        self.records.append(record)
        self._embeddings[record.id] = list(self.embedder(record.text))
        if self.path:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            with self.path.open("a", encoding="utf-8") as handle:
                handle.write(json.dumps(asdict(record), ensure_ascii=False) + "\n")
        return record

    def recall(self, query: str, *, limit: int = 5) -> list[tuple[MemoryRecord, float]]:
        if limit <= 0:
            return []
        q = list(self.embedder(query))
        now = time.time()
        sw, rw, salw, cw = self.weights
        ranked = []
        for record in self.records:
            semantic = cosine(q, self._embeddings[record.id])
            age_hours = max(0.0, (now - record.timestamp) / 3600.0)
            recency = 1.0 / (1.0 + age_hours / 24.0)
            score = sw * semantic + rw * recency + salw * record.salience + cw * record.confidence
            ranked.append((record, score))
        ranked.sort(key=lambda item: item[1], reverse=True)
        return ranked[:limit]

    def snapshot(self) -> dict[str, object]:
        return {"count": len(self.records), "path": str(self.path) if self.path else None}

    def reset(self) -> None:
        self.records.clear()
        self._embeddings.clear()

File: cosmic rust/src/memory.rs
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Clone, Debug, PartialEq)]
pub struct MemoryRecord {
    pub id: u64,
    pub text: String,
    pub timestamp: f64,
    pub salience: f64,
    pub confidence: f64,
}

#[derive(Clone, Debug)]
pub struct SemanticMemory {
    path: Option<PathBuf>,
    records: Vec<MemoryRecord>,
    embeddings: Vec<Vec<f64>>,
    dimension: usize,
    next_id: u64,
}

impl Default for SemanticMemory {
    fn default() -> Self {
        Self::new(None::<&Path>, 128).expect("in-memory semantic memory")
    }
}

impl SemanticMemory {
    pub fn new(path: Option<impl AsRef<Path>>, dimension: usize) -> Result<Self, String> {
        if dimension == 0 {
            return Err("embedding dimension must be positive".into());
        }
        let path = path.map(|value| value.as_ref().to_path_buf());
        let mut memory = Self {
            path,
            records: Vec::new(),
            embeddings: Vec::new(),
            dimension,
            next_id: 1,
        };
        memory.load()?;
        Ok(memory)
    }

    pub fn store(&mut self, text: impl Into<String>) -> Result<MemoryRecord, String> {
        self.store_scored(text, 0.5, 1.0)
    }

    pub fn store_scored(
        &mut self,
        text: impl Into<String>,
        salience: f64,
        confidence: f64,
    ) -> Result<MemoryRecord, String> {
        let text = text.into();
        if text.trim().is_empty() {
            return Err("memory text cannot be empty".into());
        }
        let record = MemoryRecord {
            id: self.next_id,
            text,
            timestamp: now(),
            salience: salience.clamp(0.0, 1.0),
            confidence: confidence.clamp(0.0, 1.0),
        };
        self.next_id += 1;
        self.embeddings.push(hashed_embedding(&record.text, self.dimension));
        self.records.push(record.clone());
        self.append(&record)?;
        Ok(record)
    }

    pub fn recall(&self, query: &str, limit: usize) -> Vec<(MemoryRecord, f64)> {
        let query_embedding = hashed_embedding(query, self.dimension);
        let current = now();
        let mut ranked: Vec<_> = self
            .records
            .iter()
            .zip(&self.embeddings)
            .map(|(record, embedding)| {
                let semantic = cosine(&query_embedding, embedding);
                let age_hours = ((current - record.timestamp).max(0.0)) / 3600.0;
                let recency = 1.0 / (1.0 + age_hours / 24.0);
                let score = 0.70 * semantic
                    + 0.10 * recency
                    + 0.15 * record.salience
                    + 0.05 * record.confidence;
                (record.clone(), score)
            })
            .collect();
        ranked.sort_by(|a, b| b.1.total_cmp(&a.1));
        ranked.truncate(limit);
        ranked
    }

    pub fn records(&self) -> &[MemoryRecord] {
        &self.records
    }

    fn append(&self, record: &MemoryRecord) -> Result<(), String> {
        let Some(path) = &self.path else {
            return Ok(());
        };
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).map_err(|error| error.to_string())?;
        }
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(path)
            .map_err(|error| error.to_string())?;
        writeln!(
            file,
            "{}\t{}\t{}\t{}\t{}",
            record.id,
            record.timestamp,
            record.salience,
            record.confidence,
            hex_encode(record.text.as_bytes())
        )
        .map_err(|error| error.to_string())
    }

    fn load(&mut self) -> Result<(), String> {
        let Some(path) = &self.path else {
            return Ok(());
        };
        if !path.exists() {
            return Ok(());
        }
        let content = fs::read_to_string(path).map_err(|error| error.to_string())?;
        for line in content.lines().filter(|line| !line.trim().is_empty()) {
            let parts: Vec<_> = line.splitn(5, '\t').collect();
            if parts.len() != 5 {
                continue;
            }
            let id = parts[0].parse::<u64>().map_err(|error| error.to_string())?;
            let timestamp = parts[1].parse::<f64>().map_err(|error| error.to_string())?;
            let salience = parts[2].parse::<f64>().map_err(|error| error.to_string())?;
            let confidence = parts[3].parse::<f64>().map_err(|error| error.to_string())?;
            let bytes = hex_decode(parts[4])?;
            let text = String::from_utf8(bytes).map_err(|error| error.to_string())?;
            let record = MemoryRecord {
                id,
                text,
                timestamp,
                salience,
                confidence,
            };
            self.next_id = self.next_id.max(id + 1);
            self.embeddings.push(hashed_embedding(&record.text, self.dimension));
            self.records.push(record);
        }
        Ok(())
    }
}

pub fn hashed_embedding(text: &str, dimension: usize) -> Vec<f64> {
    let mut vector = vec![0.0; dimension];
    for token in text
        .split(|character: char| !(character.is_ascii_alphanumeric() || character == '_'))
        .filter(|token| !token.is_empty())
    {
        let lowercase = token.to_ascii_lowercase();
        let hash = fnv1a64(lowercase.as_bytes());
        let index = (hash as usize) % dimension;
        let sign = if (hash >> 8) & 1 == 1 { 1.0 } else { -1.0 };
        vector[index] += sign;
    }
    let norm = vector.iter().map(|value| value * value).sum::<f64>().sqrt();
    if norm > 0.0 {
        for value in &mut vector {
            *value /= norm;
        }
    }
    vector
}

fn cosine(a: &[f64], b: &[f64]) -> f64 {
    let denominator = a.iter().map(|value| value * value).sum::<f64>().sqrt()
        * b.iter().map(|value| value * value).sum::<f64>().sqrt();
    if denominator == 0.0 {
        0.0
    } else {
        a.iter().zip(b).map(|(x, y)| x * y).sum::<f64>() / denominator
    }
}

fn now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_secs_f64())
        .unwrap_or(0.0)
}

fn fnv1a64(bytes: &[u8]) -> u64 {
    let mut hash = 0xcbf29ce484222325u64;
    for byte in bytes {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(0x100000001b3);
    }
    hash
}

fn hex_encode(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut output = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        output.push(HEX[(byte >> 4) as usize] as char);
        output.push(HEX[(byte & 0x0f) as usize] as char);
    }
    output
}

fn hex_decode(value: &str) -> Result<Vec<u8>, String> {
    if value.len() % 2 != 0 {
        return Err("invalid hex string".into());
    }
    let mut output = Vec::with_capacity(value.len() / 2);
    let bytes = value.as_bytes();
    for index in (0..bytes.len()).step_by(2) {
        let high = hex_value(bytes[index])?;
        let low = hex_value(bytes[index + 1])?;
        output.push((high << 4) | low);
    }
    Ok(output)
}

fn hex_value(value: u8) -> Result<u8, String> {
    match value {
        b'0'..=b'9' => Ok(value - b'0'),
        b'a'..=b'f' => Ok(value - b'a' + 10),
        b'A'..=b'F' => Ok(value - b'A' + 10),
        _ => Err("invalid hex digit".into()),
    }
}
